### ISSUE TITLE
High accuracy variance with colsample_bytree and certain seeds

### ISSUE THREAD

**Issue by psinger** [Microsoft/LightGBM#1554]
I am currently observing high variance in accuracy with using `colsample_bytree` and specific seeds. Changing the seed can change the accuracy by at least 10%. 

I am wondering how to tackle this issue. Not use feature sampling at all or maybe even use some form of bagging with different seeds?

**Comment by psinger** [comment]
@bbennett36 Currently, I have two models with different number of iterations in Python defined as follows:

`lgb.LGBMClassifier(objective="binary", seed=42, n_jobs=4, boost_from_average=False, n_estimators=100, learning_rate=0.1, colsample_bytree=0.5, max_depth=-1, min_child_samples=200, num_leaves=255, scale_pos_weight=55)`

`lgb.LGBMClassifier(objective="binary", seed=42, n_jobs=4, boost_from_average=False, n_estimators=1000, learning_rate=0.02, colsample_bytree=0.5, max_depth=-1, min_child_samples=200, num_leaves=255, scale_pos_weight=70)`

You are totally correct that the variance is much higher for the 100 iteration model with different seeds. However, it is also imminent from the 1000 iteration model, specifically for test dataset, to a lesser extend on the cross validation results.

**Comment by bbennett36** [comment]
@psinger I would guess that its just overfitting based on those parameters. Try setting num_leaves=40, min_samples_leaf=50, and instead of using scale_pos_weight try setting is_unbalance=true. That will make the weights equal to the event rate. Scale_pos_weight=70 might be too high. I don't think ive ever went over like 10.  
  
Let me know if this helps your issue.

**Comment by psinger** [comment]
@bbennett36 Thanks for help, but I have heavily unbalanced data and scaling it by 70 is not extreme at all in my case, `is_unbalance=True` nearly never works for me as it is scaling way too aggressively.

**Comment by bbennett36** [comment]
@psinger Gotcha.  I'm not sure what metric you're using but I'll usually use 'binary_logloss" when dealing with unbalanced data. This will usually make the predictions closer to the actual event rate.  
  
Also, early stopping rounds will help control overfitting too if you're not already using it

**Comment by psinger** [comment]
@bbennett36 My final goal is more in the direction of ranking. Also, early stopping stops way too early for my data at hand. I guess everything boils down to the fact that my data is quite problematic ;)

**Comment by Laurae2** [comment]
I've seen 30% accuracy variance with stratified cross-validation on a poorly made dataset, and 20% accuracy variance on a very unbalanced dataset (1:10000 ratio).

When introducing very low column / row sampling ratios, make sure to increase significantly the number of boosting iterations. For a value of 0.02, I usually recommend at least 10,000 iterations (with a learning rate low enough to not overfit before 10,000 iterations).

Another issue when introducing such ratio is instability of early stopping, in that case cranking it to 500 or more is not uncommon (also with very high regularization, boosting tends to stop itself by building empty trees).

**Comment by psinger** [comment]
@Laurae2 Thanks! What do you see as very low column / row sampling ratios. What learning rate do you propose for 10,000 iterations? 

I am also wondering regarding what metric to choose for early stopping in highly unbalanced datasets where final goal is to achieve good ranking. My current perspective on this is that something like AUC does not work properly for early stopping then, and one might need to rely on logloss with, as you said, quite high number of steps.

**Comment by Laurae2** [comment]
@psinger If we assume you use 0.3 for 100 iterations, try 0.003 (= 0.3 / (10000 / 100) ) for 10000 iterations.

If your final target is ranking, keep AUC as long as you are interested into "top N" predictions for your use case. AUC only cares about order. Otherwise, switch to logarithmic loss.

**Comment by psinger** [comment]
@Laurae2 I am actually interested into top N predictions. Thanks for elaborations.

**Comment by StrikerRUS** [comment]
@psinger @Laurae2 Can we close this?
